### PR TITLE
feat: add feature and target surface

### LINKED ISSUE #535: Contribute the-lab feature/target surface so downstream can run 100% on Limen
## Goal

`the-lab` is migrating to use Limen as the single source of truth for both data **and** features/targets. The audit below lists every feature/target currently maintained in [`the-lab/features.py`](https://github.com/Vaquum/the-lab/blob/185335ca2934c3fb1df44b3e1d06901ad325b8b4/features.py) and [`the-lab/targets.py`](https://github.com/Vaquum/the-lab/blob/185335ca2934c3fb1df44b3e1d06901ad325b8b4/targets.py) that Limen does not yet provide. Once these land, `the-lab` deletes its entire local feature/target surface and consumes everything from `limen.features`, `limen.indicators`, and `limen.targets`.

All input columns assumed below (`open`, `high`, `low`, `close`, `volume`, `maker_ratio`, `no_of_trades`, `open_liquidity`, `high_liquidity`, `low_liquidity`, `close_liquidity`, `liquidity_sum`, `maker_volume`, `maker_liquidity`, `datetime`) are already emitted by `HistoricalData.get_spot_klines` / `get_spot_dollar_klines`, so there is no upstream data dependency to resolve first.

What is **already covered** by Limen and will be dropped from `the-lab` with no Limen change required:

| `the-lab` symbol | Limen equivalent |
|---|---|
| `features.rsi(n)` | `limen.indicators.rsi_sma(period)` |
| `features.close_position(1)` (single-bar) | `limen.features.close_position` |
| `features.is_weekend` | `limen.features.calendar_time_features.is_weekend` |
| `features.nth_hour/minute/day/month/quarter` | `limen.features.calendar_time_features.{hour,minute,day_of_month,month,quarter}` (0-based; we just stop adding `+1` locally) |
| `targets.next_bar_up`, `next_bar_down` | `limen.targets.NextBarUpTarget`, `NextBarDownTarget` |
| `features.roc_*`, `log_return_*`, `realized_volatility_*`, `realized_semivolatility_down_*`, `parkinson_volatility_*`, `jump_volatility_*`, `atr_*` (the seven `*_bps_per_bar` wrappers) | covered by underlying Limen primitives `roc`, `returns`/`window_return`, `rolling_volatility`, `realized_semivariance`, `parkinson_volatility`, `jump_variation_proxy`, `atr_sma` — `the-lab` is giving up the `bps_per_bar` convention and will consume Limen's raw forms directly |

Everything below is **not** in Limen today.

---

### A.1 `ForwardVolNormalizedReturnTarget`

A predictive label: forward log-return divided by current Parkinson volatility.

Limen already ships `VolNormalizedReturnTarget`, but it is **backward-looking** (`log(close_t / close_{t-1}) / σ_{t-1}`) and therefore cannot be used as a supervised label.

Definition (one-step variant):

```
parkinson_var = (log(high) - log(low))^2 / (4 * ln 2)
σ            = sqrt( ewm_mean(parkinson_var, half_life=50, min_samples=150) )
fwd          = log(close.shift(-periods)) - log(close)
target       = (|fwd| if absolute else fwd) / σ
```

Suggested signature:

```python
ForwardVolNormalizedReturnTarget(
    train_data: pl.DataFrame,
    target_name: str,
    periods: int = 1,
    absolute: bool = False,
    halflife: int = 50,
    min_periods: int = 150,
    high_col: str = "high",
    low_col: str = "low",
    open_col: str = "open",
    close_col: str = "close",
)
```

Validation should reuse the existing Parkinson-vs-close σ sanity gate from `VolNormalizedReturnTarget`.

Reference: [`the-lab/targets.py` L69–L102](https://github.com/Vaquum/the-lab/blob/185335ca2934c3fb1df44b3e1d06901ad325b8b4/targets.py#L69-L102).

---

### B.1 `rolling_zscore`

A single primitive that subsumes 13 named z-score features in `the-lab`.

```python
def rolling_zscore(
    data: pl.DataFrame,
    column: str,
    window: int,
    *,
    transform: Literal["identity", "log1p", "abs"] = "identity",
    output_col: str | None = None,
) -> pl.DataFrame:
    x = pl.col(column)
    if transform == "log1p": x = x.log1p()
    elif transform == "abs": x = x.abs()
    mean = x.rolling_mean(window_size=window)
    std  = x.rolling_std(window_size=window)
    z    = pl.when(std == 0.0).then(None).otherwise((x - mean) / std)
    name = output_col or f"{column}_zscore_{window}"
    return data.with_columns(z.alias(name))
```

`the-lab` named wrappers that collapse to this primitive (formula = z-score of the transformed expression over `window=n`):

| Wrapper | Underlying expression | Transform |
|---|---|---|
| `liquidity_z(n)` | `liquidity_sum` | `log1p` |
| `volume_z(n)` | `volume` | `log1p` |
| `volatility_z(n)` | `parkinson_var` | `log1p` |
| `trade_size_z(n)` | `volume / max(no_of_trades, 1)` | `log1p` |
| `volume_to_liquidity_z(n)` | `volume / liquidity_sum` | `log1p` |
| `trades_z(n)` | `no_of_trades` | `log1p` |
| `roc_z(n)` | `close.pct_change()` | identity |
| `abs_price_z(n)` | `close` | identity, then `abs` after standardization |
| `maker_ratio_z(n)` / `maker_ratio_trend(n)` | `maker_ratio` | identity |
| `maker_liquidity_share_z(n)` | `maker_liquidity / liquidity_sum` | identity |
| `taker_imbalance_z(n)` | `volume * (1 - 2*maker_ratio)` | identity |
| `taker_imbalance_abs_z(n)` | same, then `abs` of the input | identity |
| `high_low_range_trend(n)` | `high - low` | identity |

References: [features.py L84–L147, L182–L188, L340–L370, L409–L425](https://github.com/Vaquum/the-lab/blob/185335ca2934c3fb1df44b3e1d06901ad325b8b4/features.py#L84-L425).

---

## C. Microstructure features (9)

All operate on columns Limen already emits.

| Function | Formula |
|---|---|
| `maker_volume_share()` | `maker_volume / volume` |
| `maker_liquidity_share()` | `maker_liquidity / liquidity_sum` |
| `maker_volume_ratio(n)` | `rolling_mean(maker_volume / volume, n)` |
| `trade_imbalance(n)` | `Σ_n maker_volume / Σ_n volume` |
| `taker_imbalance_ratio(n)` | `rolling_mean(|volume·(1 − 2·maker_ratio)| / volume, n)` |
| `trade_density(n)` | `rolling_mean(no_of_trades / volume, n)` |
| `trade_size_ratio(short_n, long_n)` | `SMA_{short_n}(volume/no_of_trades) / SMA_{long_n}(volume/no_of_trades)` |
| `liquidity_range(n)` | `rolling_mean(high_liquidity / low_liquidity, n)` |
| `liquidity_drop(n)` | `liquidity_sum_t / liquidity_sum_{t−n}` (distinct from Amihud-based `illiquidity_shock`; this is a raw LOB-liquidity ratio over a fixed lag) |

References: [features.py L166–L222, L372–L377, L393–L396, L428–L437](https://github.com/Vaquum/the-lab/blob/185335ca2934c3fb1df44b3e1d06901ad325b8b4/features.py#L166-L437).

---

## D. Structural rolling features (15)

| Function | Formula / note |
|---|---|
| `wick_proportion(n)` | `rolling_mean((range − |close−open|) / range, n)` — distinct from existing `wick_imbalance` which is asymmetric |
| `stochastic_k_abs(n)` | `|stoch_K_n − 0.5|`; uses rolling min/max of low/high over `n` |
| `volatility_ratio(short_n, long_n)` | `SMA_{short_n}(parkinson_var) / SMA_{long_n}(parkinson_var)` |
| `close_position_rolling(n)` | rolling mean of existing `close_position` — please add a `window` kwarg to `close_position` (default 1 = current behavior) |
| `distance_from_ma(n)` | `(close − SMA_n) / SMA_n` (analogous to existing `distance_from_high`/`distance_from_low`) |
| `close_ma_distance_atr(n_ma, n_atr)` | `(close − SMA_{n_ma}) / ATR_{n_atr}` where `ATR` is price-space true-range mean |
| `kaufman_efficiency_ratio(n)` | `|close_t − close_{t−n}| / Σ_{n} |close − close.shift(1)|` |
| `return_autocorrelation(n)` | rolling Pearson `corr(r_t, r_{t−1})` over `n` |
| `volume_volatility_correlation(n)` | rolling Pearson `corr(volume, parkinson_var)` over `n` |
| `return_volatility_correlation(n)` | rolling Pearson `corr(return, parkinson_var)` over `n` |
| `downside_volatility_ratio(n)` | `Σ_n r²·1{r<0} / Σ_n r²` |
| `narrow_range(n)` | `range_t / rolling_max(range, n)` |
| `volume_to_range(n)` | `rolling_mean(volume / (high − low), n)` |
| `volatility_spike(n)` | `parkinson_var_t / parkinson_var_{t−n}` — distinct from existing `volume_spike` which is `volume / rolling_mean(volume)` |
| `parkinson_vol_of_vol(n)` | `rolling_std(parkinson_var, n)` — namespace under a distinct name; Limen's existing `volatility_of_volatility` is `rolling_std(rolling_std(returns))`, both definitions are valuable |

References: [features.py L149–L271, L286–L407](https://github.com/Vaquum/the-lab/blob/185335ca2934c3fb1df44b3e1d06901ad325b8b4/features.py#L149-L407).

---

## E. Calendar gaps (3)

`calendar_time_features` and `cyclical_time_features` cover hour, minute, weekday, day_of_month, day_of_year, week_of_year, month, quarter, is_weekend. Missing:

| Function | Definition |
|---|---|
| `half_of_year` column inside `calendar_time_features` | `((month − 1) // 6) + 1` ∈ {1, 2} |
| `is_funding_hour(hours=(0, 8, 16))` | `hour.is_in(hours).cast(Int8)` — perp funding cadence indicator |
| `is_us_open_hour(hour=14)` | `(hour == hour_arg).cast(Int8)` — US cash-equity session open (14:30 UTC, hour-rounded) |

The latter two are deliberately parameterized rather than hard-coded so they generalize beyond BTC/US conventions.

References: [features.py L440–L455, L478–L479](https://github.com/Vaquum/the-lab/blob/185335ca2934c3fb1df44b3e1d06901ad325b8b4/features.py#L440-L479).

---

## Summary

| Group | Count |
|---|---|
| A. Targets | 1 |
| B. Generic utility | 1 |
| C. Microstructure features | 9 |
| D. Structural rolling features | 15 |
| E. Calendar gaps | 3 |
| **Total** | **29** |

Once these are merged, `the-lab/features.py`, `the-lab/targets.py`, and the entire local half of `the-lab/feature_registry.py` collapse to imports from `limen.*`.

### PR DESCRIPTION
**Pull request by mikkokotila**
## Summary
- Closes #535
- Keeps the new surface inside Limen's package conventions: one feature per feature module, public exports from `limen.features`, and no grouped feature bundles.
- Adds `ForwardVolNormalizedReturnTarget` and exports/docs for downstream usage.
- Routes regression coverage into the existing feature test suites by domain.
- Adds the required `CHANGELOG.md` entry, bumps package version to `3.9.0`, and records the future PR rule in `AGENTS.md` and `Claude.md`.

## Proof
- `.venv/bin/ruff check limen/features tests/test_feature_library_primitives.py tests/test_feature_library_liquidity.py tests/test_feature_library_range_volatility.py tests/test_feature_library_realized_risk.py tests/test_feature_library_seasonality.py tests/run.py` -> passed
- `.venv/bin/pytest tests/test_feature_library_primitives.py tests/test_feature_library_liquidity.py tests/test_feature_library_range_volatility.py tests/test_feature_library_realized_risk.py tests/test_feature_library_seasonality.py tests/test_targets.py` -> 60 passed
- `git diff --check` -> passed
- `.venv/bin/python -c "import tomli; tomli.load(open('pyproject.toml','rb'))"` -> passed
- `.venv/bin/python -m tests.run` -> 732 passed, 0 failed on pushed head `044f399`

### CODE REVIEW

File: limen/features/close_position.py [hunk 1]
```diff
@@ -1,16 +1,33 @@
...
+    data: pl.DataFrame,
+    window: int = 1,
+    high_col: str = 'high',
+    low_col: str = 'low',
+    close_col: str = 'close',
+    output_col: str = 'close_position',
+) -> pl.DataFrame:
 
     '''
     Compute close position within the high-low range as percentage.
```
[the remark this answers is not in the document]
Comment (zero-bang) [reply limen/features/close_position.py]: Already addressed — the new docstring (line 17) says "unit interval fraction", not "percentage". The old wording was removed in this PR.

File: limen/features/close_position.py [hunk 2]
```diff
@@ -1,16 +1,33 @@
...
+        pl.DataFrame: The input data with a new close-position column
     '''
 
-    return data.with_columns(((pl.col('close') - pl.col('low')) / (pl.col('high') - pl.col('low') + 1e-8)).alias('close_position'))
+    position = (pl.col(close_col) - pl.col(low_col)) / (pl.col(high_col) - pl.col(low_col) + 1e-8)
+    if window == 1:
+        return data.with_columns(position.alias(output_col))
+    if window <= 0:
+        raise ValueError('window must be positive')
+    return data.with_columns(position.rolling_mean(window_size=window).alias(output_col))
```
[the remark this answers is not in the document]
Comment (zero-bang) [reply limen/features/close_position.py]: Already addressed — the PR introduces an `EPSILON = 1e-10` constant at module level and uses a `pl.when(price_range.abs() <= EPSILON).then(None)` guard instead of the old inline `+ 1e-8` in the denominator.

File: tests/test_feature_library_the_lab_surface.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    identity = rolling_zscore(data, 'x', 2)
+    log_result = rolling_zscore(data, 'x', 2, transform='log1p', output_col='x_log_z')
+    flat = rolling_zscore(data, 'flat', 2)
+
+    assert identity['x_zscore_2'].to_list()[0] is None
+    assert identity['x_zscore_2'].to_list()[1:] == pytest.approx([math.sqrt(0.5), math.sqrt(0.5)])
+    assert log_result['x_log_z'].to_list()[1] == pytest.approx(math.sqrt(0.5))
+    assert flat['flat_zscore_2'].to_list()[1:] == [None, None]
+
+
```
[the remark this answers is not in the document]
Comment (zero-bang) [reply tests/test_feature_library_the_lab_surface.py]: Already covered in this test — lines 76–80 exercise the `abs` path (`abs_result`) and the invalid-transform error (`pytest.raises(ValueError, match='transform must be one of')`).

**Review by zero-bang — APPROVED** [review]
Thorough review — clean PR.

**Correctness**: All feature functions use proper null-safe patterns (`_safe_divide`, `pl.when` guards). The `ForwardVolNormalizedReturnTarget` correctly uses current sigma (no shift) with forward return, which is the right choice for a predictive label. The `close_position` refactor from epsilon-in-denominator to null-on-zero-range is a strict improvement. All new parameters default to backward-compatible values.

**Tests**: 5 feature tests + 2 target tests cover every new function, including edge cases (zero-std nulls, alternating returns for autocorrelation, constant-variance series). The existing `test_close_position_uses_intra_candle_range` test continues to pass under the new signature since all its SAMPLE_OHLCV ranges are positive.

**Exports & registration**: All 27 new feature functions and 1 new target class are imported, listed in `__all__`, and all 7 new test functions are registered in `tests/run.py`. Alphabetical ordering is maintained throughout.

**Docs**: Features.md, Targets.md, and the features README are all updated and consistent with the implementation.

**Copilot comments**: All three were already addressed in the final diff — replied to each thread.
